### PR TITLE
Replace exporter dependencies with standardised Open XML SDK

### LINKED ISSUE #42: As a developer, I want to use a standardised Office SDK, so that I don't need multiple overlapping libraries to generate Office documents
At this moment there are multiple libraries used to create export files. 
[Open_XML-SDK](https://github.com/OfficeDev/Open-XML-SDK) has support for, word documents and spreadsheets. It is an open-source lib created by the Office developer team from Microsoft.

This is for now maybe a nice to have.  But with more features, this lib has. It can become a need when more features are required

### PR DESCRIPTION
**Pull request by NealGeilen**
A proof of concept is included that can be tested on multiple operating systems.

### CODE REVIEW

**Comment by NealGeilen** [comment]
Sorry for the Git mess. Can't really find what went wrong here.

**Comment by NealGeilen** [comment]
@Typiqally Could you test the word export feature on a Mac with this branch?
The code is a mess but i just wanted to know if it works.

**Comment by Typiqally** [comment]
> Sorry for the Git mess. Can't really find what went wrong here.

Maybe you didn't update your master branch on local?

**Comment by NealGeilen** [comment]
> > Sorry for the Git mess. Can't really find what went wrong here.
> 
> Maybe you didn't update your master branch on local?

I thought I did, But it is possible. idk

**Comment by Typiqally** [comment]
I get the following errors on macOS

```
Error 1
Description: The element has unexpected child element 'http://schemas.openxmlformats.org/wordprocessingml/2006/main:tr'. List of possible elements expected: <http://schemas.openxmlformats.org/wordprocessingml/2006/main:bookmarkEnd>,<http://schemas.openxmlformats.org/wordprocessingml/2006/main:bookmarkStart>,<http://schemas.openxmlformats.org/wordprocessingml/2006/main:commentRangeEnd>,<http://schemas.openxmlformats.org/wordprocessingml/2006/main:commentRangeStart>,<http://schemas.openxmlformats.org/wordprocessingml/2006/main:customXmlDelRangeEnd>,<http://schemas.openxmlformats.org/wordprocessingml/2006/main:customXmlDelRangeStart>,<http://schemas.openxmlformats.org/wordprocessingml/2006/main:customXmlInsRangeEnd>,<http://schemas.openxmlformats.org/wordprocessingml/2006/main:customXmlInsRangeStart>,<http://schemas.openxmlformats.org/wordprocessingml/2006/main:customXmlMoveFromRangeEnd>,<http://schemas.openxmlformats.org/wordprocessingml/2006/main:customXmlMoveFromRangeStart>,<http://schemas.openxmlformats.org/wordprocessingml/2006/main:customXmlMoveToRangeEnd>,<http://schemas.openxmlformats.org/wordprocessingml/2006/main:customXmlMoveToRangeStart>,<http://schemas.openxmlformats.org/wordprocessingml/2006/main:moveFromRangeEnd>,<http://schemas.openxmlformats.org/wordprocessingml/2006/main:moveFromRangeStart>,<http://schemas.openxmlformats.org/wordprocessingml/2006/main:moveToRangeEnd>,<http://schemas.openxmlformats.org/wordprocessingml/2006/main:moveToRangeStart>.
ErrorType: Schema
Node: DocumentFormat.OpenXml.Wordprocessing.Table
Path: /w:document[1]/w:body[1]/w:tbl[3]
Part: /word/document.xml
-------------------------------------------
Error 2
Description: The element has unexpected child element 'http://schemas.openxmlformats.org/wordprocessingml/2006/main:tr'. List of possible elements expected: <http://schemas.openxmlformats.org/wordprocessingml/2006/main:bookmarkEnd>,<http://schemas.openxmlformats.org/wordprocessingml/2006/main:bookmarkStart>,<http://schemas.openxmlformats.org/wordprocessingml/2006/main:commentRangeEnd>,<http://schemas.openxmlformats.org/wordprocessingml/2006/main:commentRangeStart>,<http://schemas.openxmlformats.org/wordprocessingml/2006/main:customXmlDelRangeEnd>,<http://schemas.openxmlformats.org/wordprocessingml/2006/main:customXmlDelRangeStart>,<http://schemas.openxmlformats.org/wordprocessingml/2006/main:customXmlInsRangeEnd>,<http://schemas.openxmlformats.org/wordprocessingml/2006/main:customXmlInsRangeStart>,<http://schemas.openxmlformats.org/wordprocessingml/2006/main:customXmlMoveFromRangeEnd>,<http://schemas.openxmlformats.org/wordprocessingml/2006/main:customXmlMoveFromRangeStart>,<http://schemas.openxmlformats.org/wordprocessingml/2006/main:customXmlMoveToRangeEnd>,<http://schemas.openxmlformats.org/wordprocessingml/2006/main:customXmlMoveToRangeStart>,<http://schemas.openxmlformats.org/wordprocessingml/2006/main:moveFromRangeEnd>,<http://schemas.openxmlformats.org/wordprocessingml/2006/main:moveFromRangeStart>,<http://schemas.openxmlformats.org/wordprocessingml/2006/main:moveToRangeEnd>,<http://schemas.openxmlformats.org/wordprocessingml/2006/main:moveToRangeStart>.
ErrorType: Schema
Node: DocumentFormat.OpenXml.Wordprocessing.Table
Path: /w:document[1]/w:body[1]/w:tbl[2]
Part: /word/document.xml
-------------------------------------------
Error 3
Description: The element has unexpected child element 'http://schemas.openxmlformats.org/wordprocessingml/2006/main:tr'. List of possible elements expected: <http://schemas.openxmlformats.org/wordprocessingml/2006/main:bookmarkEnd>,<http://schemas.openxmlformats.org/wordprocessingml/2006/main:bookmarkStart>,<http://schemas.openxmlformats.org/wordprocessingml/2006/main:commentRangeEnd>,<http://schemas.openxmlformats.org/wordprocessingml/2006/main:commentRangeStart>,<http://schemas.openxmlformats.org/wordprocessingml/2006/main:customXmlDelRangeEnd>,<http://schemas.openxmlformats.org/wordprocessingml/2006/main:customXmlDelRangeStart>,<http://schemas.openxmlformats.org/wordprocessingml/2006/main:customXmlInsRangeEnd>,<http://schemas.openxmlformats.org/wordprocessingml/2006/main:customXmlInsRangeStart>,<http://schemas.openxmlformats.org/wordprocessingml/2006/main:customXmlMoveFromRangeEnd>,<http://schemas.openxmlformats.org/wordprocessingml/2006/main:customXmlMoveFromRangeStart>,<http://schemas.openxmlformats.org/wordprocessingml/2006/main:customXmlMoveToRangeEnd>,<http://schemas.openxmlformats.org/wordprocessingml/2006/main:customXmlMoveToRangeStart>,<http://schemas.openxmlformats.org/wordprocessingml/2006/main:moveFromRangeEnd>,<http://schemas.openxmlformats.org/wordprocessingml/2006/main:moveFromRangeStart>,<http://schemas.openxmlformats.org/wordprocessingml/2006/main:moveToRangeEnd>,<http://schemas.openxmlformats.org/wordprocessingml/2006/main:moveToRangeStart>.
ErrorType: Schema
Node: DocumentFormat.OpenXml.Wordprocessing.Table
Path: /w:document[1]/w:body[1]/w:tbl[1]
Part: /word/document.xml
-------------------------------------------
count=3
Error 1
Description: The element has unexpected child element 'http://schemas.openxmlformats.org/wordprocessingml/2006/main:tr'. List of possible elements expected: <http://schemas.openxmlformats.org/wordprocessingml/2006/main:bookmarkEnd>,<http://schemas.openxmlformats.org/wordprocessingml/2006/main:bookmarkStart>,<http://schemas.openxmlformats.org/wordprocessingml/2006/main:commentRangeEnd>,<http://schemas.openxmlformats.org/wordprocessingml/2006/main:commentRangeStart>,<http://schemas.openxmlformats.org/wordprocessingml/2006/main:customXmlDelRangeEnd>,<http://schemas.openxmlformats.org/wordprocessingml/2006/main:customXmlDelRangeStart>,<http://schemas.openxmlformats.org/wordprocessingml/2006/main:customXmlInsRangeEnd>,<http://schemas.openxmlformats.org/wordprocessingml/2006/main:customXmlInsRangeStart>,<http://schemas.openxmlformats.org/wordprocessingml/2006/main:customXmlMoveFromRangeEnd>,<http://schemas.openxmlformats.org/wordprocessingml/2006/main:customXmlMoveFromRangeStart>,<http://schemas.openxmlformats.org/wordprocessingml/2006/main:customXmlMoveToRangeEnd>,<http://schemas.openxmlformats.org/wordprocessingml/2006/main:customXmlMoveToRangeStart>,<http://schemas.openxmlformats.org/wordprocessingml/2006/main:moveFromRangeEnd>,<http://schemas.openxmlformats.org/wordprocessingml/2006/main:moveFromRangeStart>,<http://schemas.openxmlformats.org/wordprocessingml/2006/main:moveToRangeEnd>,<http://schemas.openxmlformats.org/wordprocessingml/2006/main:moveToRangeStart>.
ErrorType: Schema
Node: DocumentFormat.OpenXml.Wordprocessing.Table
Path: /w:document[1]/w:body[1]/w:tbl[3]
Part: /word/document.xml
-------------------------------------------
Error 2
Description: The element has unexpected child element 'http://schemas.openxmlformats.org/wordprocessingml/2006/main:tr'. List of possible elements expected: <http://schemas.openxmlformats.org/wordprocessingml/2006/main:bookmarkEnd>,<http://schemas.openxmlformats.org/wordprocessingml/2006/main:bookmarkStart>,<http://schemas.openxmlformats.org/wordprocessingml/2006/main:commentRangeEnd>,<http://schemas.openxmlformats.org/wordprocessingml/2006/main:commentRangeStart>,<http://schemas.openxmlformats.org/wordprocessingml/2006/main:customXmlDelRangeEnd>,<http://schemas.openxmlformats.org/wordprocessingml/2006/main:customXmlDelRangeStart>,<http://schemas.openxmlformats.org/wordprocessingml/2006/main:customXmlInsRangeEnd>,<http://schemas.openxmlformats.org/wordprocessingml/2006/main:customXmlInsRangeStart>,<http://schemas.openxmlformats.org/wordprocessingml/2006/main:customXmlMoveFromRangeEnd>,<http://schemas.openxmlformats.org/wordprocessingml/2006/main:customXmlMoveFromRangeStart>,<http://schemas.openxmlformats.org/wordprocessingml/2006/main:customXmlMoveToRangeEnd>,<http://schemas.openxmlformats.org/wordprocessingml/2006/main:customXmlMoveToRangeStart>,<http://schemas.openxmlformats.org/wordprocessingml/2006/main:moveFromRangeEnd>,<http://schemas.openxmlformats.org/wordprocessingml/2006/main:moveFromRangeStart>,<http://schemas.openxmlformats.org/wordprocessingml/2006/main:moveToRangeEnd>,<http://schemas.openxmlformats.org/wordprocessingml/2006/main:moveToRangeStart>.
ErrorType: Schema
Node: DocumentFormat.OpenXml.Wordprocessing.Table
Path: /w:document[1]/w:body[1]/w:tbl[2]
Part: /word/document.xml
-------------------------------------------
Error 3
Description: The element has unexpected child element 'http://schemas.openxmlformats.org/wordprocessingml/2006/main:tr'. List of possible elements expected: <http://schemas.openxmlformats.org/wordprocessingml/2006/main:bookmarkEnd>,<http://schemas.openxmlformats.org/wordprocessingml/2006/main:bookmarkStart>,<http://schemas.openxmlformats.org/wordprocessingml/2006/main:commentRangeEnd>,<http://schemas.openxmlformats.org/wordprocessingml/2006/main:commentRangeStart>,<http://schemas.openxmlformats.org/wordprocessingml/2006/main:customXmlDelRangeEnd>,<http://schemas.openxmlformats.org/wordprocessingml/2006/main:customXmlDelRangeStart>,<http://schemas.openxmlformats.org/wordprocessingml/2006/main:customXmlInsRangeEnd>,<http://schemas.openxmlformats.org/wordprocessingml/2006/main:customXmlInsRangeStart>,<http://schemas.openxmlformats.org/wordprocessingml/2006/main:customXmlMoveFromRangeEnd>,<http://schemas.openxmlformats.org/wordprocessingml/2006/main:customXmlMoveFromRangeStart>,<http://schemas.openxmlformats.org/wordprocessingml/2006/main:customXmlMoveToRangeEnd>,<http://schemas.openxmlformats.org/wordprocessingml/2006/main:customXmlMoveToRangeStart>,<http://schemas.openxmlformats.org/wordprocessingml/2006/main:moveFromRangeEnd>,<http://schemas.openxmlformats.org/wordprocessingml/2006/main:moveFromRangeStart>,<http://schemas.openxmlformats.org/wordprocessingml/2006/main:moveToRangeEnd>,<http://schemas.openxmlformats.org/wordprocessingml/2006/main:moveToRangeStart>.
ErrorType: Schema
Node: DocumentFormat.OpenXml.Wordprocessing.Table
Path: /w:document[1]/w:body[1]/w:tbl[1]
Part: /word/document.xml
-------------------------------------------
count=3
```

I do have a document but it is filled with gibberish

**Comment by NealGeilen** [comment]
Could you share the gibberish? 

I have the same errors, the table still has its headers on windows. (The goal of POC)

**Comment by Typiqally** [comment]
[Epsilon-Export-29112022130627.docx](https://github.com/Typiqally/epsilon/files/10112963/Epsilon-Export-29112022130627.docx)

**Comment by NealGeilen** [comment]
> [Epsilon-Export-29112022130627.docx](https://github.com/Typiqally/epsilon/files/10112963/Epsilon-Export-29112022130627.docx)

So, At my side its correct 
image


You are opening the file with word?

**Comment by Typiqally** [comment]
I'm opening the file with [Pages](https://www.apple.com/pages/), the standard processor for Word documents on macOS.

**Comment by NealGeilen** [comment]
> I'm opening the file with [Pages](https://www.apple.com/pages/), the standard processor for Word documents on macOS.

Could you try again with the latest update? I implemented an example script from the Microsoft docs now. 

The issue now is that the file can be generated by a mac. It only does not show it correctly in Pages. but it has support for the format. And why it won't show is something I don't follow.  Word can read, Google Docs can read it. and my mobile can read the format. I can't find out why Pages won't accept it. 

(and yes there are still errors when creating the file. regardless it works on the beforementioned platforms)

**Comment by Typiqally** [comment]
Same issue, see screenshot:

Screenshot 2022-11-29 at 19 25 20

Maybe specifying a set width for the columns will work?

**Comment by NealGeilen** [comment]
> Same issue, see screenshot:
> 
> Screenshot 2022-11-29 at 19 25 20
> 
> Maybe specifying a set width for the columns will work?

Yes. That's indeed something that I skipped for a POC. 
Do we need to test this also for Linux? or is this good enough to continue development? 

For me this is good enough and a clear sign that it will work.

**Comment by Typiqally** [comment]
> > Same issue, see screenshot:
> > Screenshot 2022-11-29 at 19 25 20
> > Maybe specifying a set width for the columns will work?
> 
> Yes. That's indeed something that I skipped for a POC. Do we need to test this also for Linux? or is this good enough to continue development?
> 
> For me this is good enough and a clear sign that it will work.

I feel like if it works for macOS then it will probably work on Linux as well since they are both based on Unix. Unless they are using very specific Linux instructions or packages for some reason, but I doubt that since it is really just a fancy XML generator.

**Comment by NealGeilen** [comment]
@Typiqally The PR is now complete and up to date with the main branch.

File: Epsilon/Export/Exporters/WordModuleExporter.cs [hunk 1]
```diff
@@ -16,44 +17,109 @@ public WordModuleExporter(IOptions<ExportOptions> options)
...
+    {
+        var tc = new TableCell();
+        tc.Append(CreateText(text));
+        tc.Append(new TableCellProperties(
+            new TableCellWidth { Type = TableWidthUnitValues.Auto }));
+        return tc;
+    }
+
+
+    private static TableProperties GetTableProperties()
```
Comment (Typiqally) [inline Epsilon/Export/Exporters/WordModuleExporter.cs]: Could be a constant

File: Epsilon/Export/Exporters/WordModuleExporter.cs [hunk 2]
```diff
@@ -16,44 +17,109 @@ public WordModuleExporter(IOptions<ExportOptions> options)
...
-        var link = document.AddHyperlink(Constants.ProjectName, Constants.RepositoryUri);
-
-        document.Footers.Odd
-            .InsertParagraph("Created with ")
-            .AppendHyperlink(link).Color(Color.Blue).UnderlineStyle(UnderlineStyle.singleLine);
-        
         foreach (var module in data.CourseModules)
         {
-            var table = document.AddTable(1, 3);
+            Table table = new Table();
```
Comment (Typiqally) [inline Epsilon/Export/Exporters/WordModuleExporter.cs]: We are using 'var' everywhere, so here should be the same

File: Epsilon/Export/Exporters/WordModuleExporter.cs [hunk 3]
```diff
@@ -16,44 +17,109 @@ public WordModuleExporter(IOptions<ExportOptions> options)
...
-            table.Rows[0].Cells[1].Paragraphs[0].Append("Assignment(s)");
-            table.Rows[0].Cells[2].Paragraphs[0].Append("Score");
+            table.AppendChild<TableProperties>(GetTableProperties());
+            table.Append(AddHeader());
 
             foreach (var kpi in module.Kpis)
             {
-                var row = table.InsertRow();
-                row.Cells[0].Paragraphs[0].Append(kpi.Name);
+                TableRow row = new TableRow();
```
Comment (Typiqally) [inline Epsilon/Export/Exporters/WordModuleExporter.cs]: Use 'var'

File: Epsilon/Export/Exporters/ExcelModuleExporter.cs [hunk 4]
```diff
@@ -16,36 +19,124 @@ public ExcelModuleExporter(IOptions<ExportOptions> options)
...
+                {
+                    if (String.Compare(cell.CellReference.Value, cellReference, StringComparison.OrdinalIgnoreCase) > 0)
+                    {
+                        refCell = cell;
+                        break;
+                    }
+                }
+            }
+
+            Cell newCell = new Cell() { CellReference = cellReference };
```
Comment (Typiqally) [inline Epsilon/Export/Exporters/ExcelModuleExporter.cs]: We are using 'var' everywhere, so here should be the same

File: Epsilon/Export/Exporters/ExcelModuleExporter.cs [hunk 5]
```diff
@@ -16,36 +19,124 @@ public ExcelModuleExporter(IOptions<ExportOptions> options)
...
+        uint sheetId = 1;
+        if (sheets.Elements<Sheet>().Any())
+        {
+            sheetId = sheets.Elements<Sheet>().Select(s => s.SheetId!.Value).Max() + 1;
+        }
 
-            worksheet.Cells.ColumnWidth[0, 0] = 500;
-            worksheet.Cells.ColumnWidth[0, 1] = 8000;
-            worksheet.Cells.ColumnWidth[0, 2] = 8000;
+        Sheet sheet = new Sheet() { Id = relationshipId, SheetId = sheetId, Name = module.Name };
```
Comment (Typiqally) [inline Epsilon/Export/Exporters/ExcelModuleExporter.cs]: We are using 'var' everywhere, so here should be the same

File: Epsilon/Export/Exporters/ExcelModuleExporter.cs [hunk 6]
```diff
@@ -16,36 +19,124 @@ public ExcelModuleExporter(IOptions<ExportOptions> options)
...
+        Cell cell = InsertCellInWorksheet(columnName, rowIndex, worksheetPart);
+        cell.CellValue = new CellValue(content);
+        cell.DataType = CellValues.String;
+        return cell;
+    }
+
+    // Given a WorkbookPart, inserts a new worksheet.
+    private static WorksheetPart InsertWorksheet(CourseModule module, WorkbookPart workbookPart)
+    {
+        WorksheetPart newWorksheetPart = workbookPart.AddNewPart<WorksheetPart>();
```
Comment (Typiqally) [inline Epsilon/Export/Exporters/ExcelModuleExporter.cs]: We are using 'var' everywhere, so here should be the same

File: Epsilon/Export/Exporters/ExcelModuleExporter.cs [hunk 7]
```diff
@@ -16,36 +19,124 @@ public ExcelModuleExporter(IOptions<ExportOptions> options)
...
             }
+        }
+
+        workbookpart.Workbook.Save();
+        spreadsheetDocument.Close();
+    }
+
+    private static Cell CreateTextCell(string content, string columnName, uint rowIndex, WorksheetPart worksheetPart)
+    {
+        Cell cell = InsertCellInWorksheet(columnName, rowIndex, worksheetPart);
```
Comment (Typiqally) [inline Epsilon/Export/Exporters/ExcelModuleExporter.cs]: We are using 'var' everywhere, so here should be the same

File: Epsilon/Export/Exporters/ExcelModuleExporter.cs [hunk 8]
```diff
@@ -16,36 +19,124 @@ public ExcelModuleExporter(IOptions<ExportOptions> options)
 
     public IEnumerable<string> Formats { get; } = new[] { "xls", "xlsx", "excel" };
 
-    public async Task Export(ExportData data, string format)
+    public void Export(ExportData data, string format)
     {
-        var workbook = new Workbook();
+        using SpreadsheetDocument spreadsheetDocument =
+            SpreadsheetDocument.Create($"{_options.FormattedOutputName}.xlsx", SpreadsheetDocumentType.Workbook);
+        WorkbookPart workbookpart = spreadsheetDocument.AddWorkbookPart();
```
Comment (Typiqally) [inline Epsilon/Export/Exporters/ExcelModuleExporter.cs]: We are using 'var' everywhere, so here should be the same

File: Epsilon/Export/Exporters/ExcelModuleExporter.cs [hunk 9]
```diff
@@ -16,36 +19,124 @@ public ExcelModuleExporter(IOptions<ExportOptions> options)
...
+        workbookpart.Workbook = new Workbook();
+
+        // Add Sheets to the Workbook.
+        spreadsheetDocument.WorkbookPart!.Workbook.AppendChild<Sheets>(new Sheets());
 
         foreach (var module in data.CourseModules)
         {
-            var worksheet = new Worksheet(module.Name);
-            var kpis = module.Kpis;
+            WorksheetPart worksheetPart = InsertWorksheet(module, workbookpart);
```
Comment (Typiqally) [inline Epsilon/Export/Exporters/ExcelModuleExporter.cs]: We are using 'var' everywhere, so here should be the same

File: Epsilon/Export/ExportDataPackager.cs [hunk 10]
```diff
@@ -27,14 +27,15 @@ await foreach (var item in data.Where(m => m.Collection.OutcomeResults.Any()))
...
-                    .Select(o => o.Link.Assignment);
+                    .Select(o => o.Link.Assignment).ToArray();
 
                 if (assignmentIds.Any())
                 {
                     var assignments = assignmentIds
                         .Select(assignmentId => new CourseAssignment
                         {
-                            Name = alignments[assignmentId].Name + " | " + alignments[assignmentId].Url,
+                            Name = alignments[assignmentId!].Name,
```
Comment (Typiqally) [inline Epsilon/Export/ExportDataPackager.cs:37]: We should start thinking about nullability here, see ! symbol

**Review by Typiqally — CHANGES_REQUESTED** [review]

File: Epsilon/Export/Exporters/WordModuleExporter.cs (hunk 1 above)
Comment (NealGeilen) [reply Epsilon/Export/Exporters/WordModuleExporter.cs]: I can't get this fixed without causing an error when running the application.

File: Epsilon/Export/Exporters/ExcelModuleExporter.cs [hunk 11]
```diff
@@ -16,36 +19,124 @@
...
+            {
+                if (cell.CellReference?.Value?.Length == cellReference.Length)
+                {
+                    if (String.Compare(cell.CellReference.Value, cellReference, StringComparison.OrdinalIgnoreCase) > 0)
+                    {
+                        refCell = cell;
+                        break;
+                    }
+                }
+            }
```
[the remark this answers is not in the document]
Comment (Typiqally) [reply Epsilon/Export/Exporters/ExcelModuleExporter.cs]: I think this would make it a bit cleaner

**Review by Typiqally — APPROVED** [review]

**Review by Typiqally — APPROVED** [review]

**Pull request merged**